Implement UnmanagedCallersOnly conversions support

333fred opened the pull request:
We now support understanding the calling convention types specified in UnmanagedCallersOnlyAttribute when performing overload resolution on an address of a method group. For the purposes of such resolution, we use the rules from the function pointers spec https://github.com/dotnet/csharplang/blob/master/proposals/csharp-9.0/function-pointers.md#systemruntimeinteropservicesunmanagedcallersonlyattribute.

Fixes https://github.com/dotnet/roslyn/issues/47858.

333fred commented:
/cc @jkoritzinsky @jkotas @tannergooding @elinor-fung, this is the PR you've been waiting for :)

333fred commented:
@dotnet/roslyn-compiler for review

333fred commented:
@dotnet/roslyn-compiler for review please.

cston commented on src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/CallingConventionInfo.cs:
```diff
@@ -0,0 +1,20 @@
+#nullable enable
```
># [](start = 0, length = 1)

Missing license.

cston commented on src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/CallingConventionInfo.cs:
```diff
@@ -0,0 +1,20 @@
+{
+    internal readonly struct CallingConventionInfo
+    {
+        internal Cci.CallingConvention CallKind { get; }
```
>{ get; } [](start = 48, length = 8)

Consider using `readonly` fields rather than properties.

cston commented on src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs:
```diff
@@ -431,28 +432,116 @@ private static bool OverloadResolutionResultIsValid<TMember>(ArrayBuilder<Member
+                    Debug.Assert(unmanagedCallersOnlyData.CallingConventionTypes.All(
+                        u => FunctionPointerTypeSymbol.IsCallingConventionModifier((NamedTypeSymbol)u)));
+
+                    // If there is only one type in the unmanagedData, and that type is one of the 4 special names known to the
```
>unmanagedData [](start = 56, length = 13)

"unmanagedCallersOnlyData"

cston commented on src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs:
```diff
@@ -431,28 +432,116 @@ private static bool OverloadResolutionResultIsValid<TMember>(ArrayBuilder<Member
-                        results[i] = new MemberResolutionResult<TMember>(
-                            result.Member, result.LeastOverriddenMember,
-                            MemberAnalysisResult.WrongCallingConvention());
+                        return;
```
>return [](start = 24, length = 6)

Why return here rather than processing remaining results? Same question for `return;` statements in `switch` below.

cston commented on src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs:
```diff
@@ -431,28 +432,116 @@ private static bool OverloadResolutionResultIsValid<TMember>(ArrayBuilder<Member
+
+                    foreach (var expectedModifier in expectedConvention.CallingConventionTypes)
+                    {
+                        if (!unmanagedCallersOnlyData.CallingConventionTypes.Contains(((CSharpCustomModifier)expectedModifier).ModifierSymbol))
```
Consider using:
```
if (expectedConvention.CallingConventionTypes.Any(
    (expectedModifier, types) => !types.Contains((CSharpCustomModifier)expectedModifier).ModiferSymbol),
    unmanagedCallersOnlyData.CallingConventionTypes))
{
    ...
}
```

333fred replied:
We don't have this extension method on ImmutableHashSet today, but I can add one.

---
In reply to: [488848549](https://github.com/dotnet/roslyn/pull/47651#discussion_r488848549) [](ancestors = 488848549)

cston replied:
When I suggested using `Any()`, I assumed `CallingConventionTypes` was an `ImmutableArray<>`. Sorry for the confusion. If we don't already have such an extension method, it's probably not worth adding one.

---
In reply to: [490551337](https://github.com/dotnet/roslyn/pull/47651#discussion_r490551337) [](ancestors = 490551337,488848549)

333fred replied:
Yeah, it has to be a hash set because we don't consider duplicates of the same attribute.

---
In reply to: [490558994](https://github.com/dotnet/roslyn/pull/47651#discussion_r490558994) [](ancestors = 490558994,490551337,488848549)

cston commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbolWithAttributes.cs:
```diff
@@ -433,8 +439,17 @@ internal sealed override ObsoleteAttributeData ObsoleteAttributeData
+                    if (lateData is { UnmanagedCallersOnlyAttributeData: not null })
+                    {
+                        // We can't verify the symmetric case here. Error conditions (such as if a bad expression was provided to the array initializer)
+                        // can cause the attribute to be entirely unbound during regular attribute binding. Early binding doesn't know that though, so
```
>attribute to be entirely unbound [](start = 41, length = 32)

"to be unbound" sounds like binding is done, then undone. Perhaps "can leave the attribute unbound" or "can cause the attribute to be skipped".

cston commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbolWithAttributes.cs:
```diff
@@ -433,8 +439,17 @@ internal sealed override ObsoleteAttributeData ObsoleteAttributeData
 #if DEBUG // Can remove ifdefs and replace with Conditional after https://github.com/dotnet/roslyn/issues/47463 is fixed
                 static void verifyDataConsistent(CommonMethodEarlyWellKnownAttributeData? earlyData, MethodWellKnownAttributeData? lateData)
```
>verifyDataConsistent [](start = 28, length = 20)

Consider inlining.

333fred replied:
I'd prefer to leave it as is, I feel that it's more readable, and it would still need `#if DEBUG`s for the `if` blocks.

---
In reply to: [488853166](https://github.com/dotnet/roslyn/pull/47651#discussion_r488853166) [](ancestors = 488853166)

cston commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -34,9 +34,19 @@ public class CodeGenFunctionPointersTests : CSharpTestBase
             bool overrideUnmanagedSupport = false)
+            => CompileAndVerifyFunctionPointers(new[] { source }, references, symbolValidator, expectedOutput, targetFramework, options, overrideUnmanagedSupport);
+
+        private CompilationVerifier CompileAndVerifyFunctionPointers(
```
>CompilationVerifier CompileAndVerifyFunctionPointers [](start = 16, length = 52)

Does these two overloads differ by `source` only? If so, consider using one overload with `CSharpTestSource source` parameter instead.

cston commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -53,12 +63,17 @@ private CompilationVerifier CompileAndVerifyFunctionPointersWithIl(string source
+            return CreateCompilation(source, references: references, options: options ?? TestOptions.UnsafeReleaseDll, parseOptions: TestOptions.Regular9);
+        }
+
+        private static CSharpCompilation CreateCompilationWithFunctionPointers(string[] source, IEnumerable<MetadataReference>? references = null, CSharpCompilationOptions? options = null)
```
>string[] source [](start = 79, length = 15)

`CSharpTestSource source`

cston commented on src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs:
```diff
@@ -431,28 +432,116 @@ private static bool OverloadResolutionResultIsValid<TMember>(ArrayBuilder<Member
+                        || ReferenceEquals(unmanagedCallersOnlyData, UnmanagedCallersOnlyAttributeData.Uninitialized))
+                    {
+                        // If we're at a location where the unmanaged data has actually not yet been bound, it cannot be
+                        // valid code anyway, as the only place that can occur is when we're actually doing an address of
```
>doing [](start = 102, length = 5)

"binding"?

AlekseyTs replied:
This assumption doesn't feel right. I think it is quite possible to attempt to access a member for which we haven't bound attributes.

cston commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -9812,6 +9851,311 @@ public UnmanagedCallersOnlyAttribute()
+
+        [Theory]
+        [InlineData("", 1)]
+        [InlineData("CallConvs = new System.Type[0]", 1)]
```
>CallConvs = new System.Type[0]" [](start = 21, length = 31)

Consider testing "CallingConvs = null", "CallingConvs = new Type[] { null }"

333fred replied:
Will do the former. The latter is covered by `UnmanagedCallersOnlyCallConvNull`.

---
In reply to: [488866675](https://github.com/dotnet/roslyn/pull/47651#discussion_r488866675) [](ancestors = 488866675)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbolWithAttributes.cs:
```diff
@@ -411,19 +411,25 @@ internal sealed override ObsoleteAttributeData ObsoleteAttributeData
                 {
-                    var data = (CommonMethodEarlyWellKnownAttributeData?)lazyCustomAttributesBag.EarlyDecodedWellKnownAttributeData;
-                    return data?.UnmanagedCallersOnlyAttributePresent == true
+                    return earlyData?.UnmanagedCallersOnlyAttributePresent == true
```
>UnmanagedCallersOnlyAttributePresent [](start = 38, length = 36)

Unless VB uses this member we should move it to C# specific type.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbolWithAttributes.cs:
```diff
@@ -411,19 +411,25 @@ internal sealed override ObsoleteAttributeData ObsoleteAttributeData
                     return UnmanagedCallersOnlyAttributeData.Uninitialized;
                 }
 
+                var earlyData = (CommonMethodEarlyWellKnownAttributeData?)lazyCustomAttributesBag.EarlyDecodedWellKnownAttributeData;
```
>EarlyDecodedWellKnownAttributeData [](start = 98, length = 34)

It doesn't feel safe to access ```EarlyDecodedWellKnownAttributeData``` without making sure that ```IsEarlyDecodedWellKnownAttributeDataComputed``` is true first

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbolWithAttributes.cs:
```diff
@@ -411,19 +411,25 @@ internal sealed override ObsoleteAttributeData ObsoleteAttributeData
                 {
-                    var data = (MethodWellKnownAttributeData?)lazyCustomAttributesBag.DecodedWellKnownAttributeData;
+                    var lateData = (MethodWellKnownAttributeData?)lazyCustomAttributesBag.DecodedWellKnownAttributeData;
+                    if (earlyData is { UnmanagedCallersOnlyAttributePresent: true } && lateData is null or { UnmanagedCallersOnlyAttributeData: null })
```
>earlyData [](start = 24, length = 9)

Is there a guarantee that early data are always computed when all attributes are decoded? It feels like this check for both early and late data is unnecessary complicated. Why can't we just rely on late data?

333fred replied:
>  Why can't we just rely on late data? 

If the attribute body is malformed (see `UnmanagedCallersOnly_BadExpressionInArguments`) for an example, we never attempt to bind the attribute at all. So the early data records the presence of an unmanaged callers only attribute, but the late data contains no such info. Now, we could not take that into account here, and the consequence would be that methods that convert such a method to a function pointer would be treated as having a default managed calling convention, as opposed to having a default unmanaged calling convention (which is what this code will do). I felt that the approach that would feel the best for users is to treat it as default unmanaged, rather than default managed, but I could revise it to return null in that case if you think it's too complicated.

---
In reply to: [488957958](https://github.com/dotnet/roslyn/pull/47651#discussion_r488957958) [](ancestors = 488957958)

AlekseyTs replied:
> I could revise it to return null in that case if you think it's too complicated.

I think this is too much trouble for little benefit. I think it is our usual approach to ignore malformed well-known attributes as if they were not applied. I think we should follow the same approach with this attribute.

---
In reply to: [490554943](https://github.com/dotnet/roslyn/pull/47651#discussion_r490554943) [](ancestors = 490554943,488957958)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs:
```diff
@@ -431,28 +432,116 @@ private static bool OverloadResolutionResultIsValid<TMember>(ArrayBuilder<Member
+                    if (unmanagedCallersOnlyData is null)
+                    {
+                        if (!member.CallingConvention.IsCallingConvention(expectedConvention.CallKind)
+                            || !expectedConvention.CallingConventionTypes.IsEmpty)
```
>!expectedConvention.CallingConventionTypes.IsEmpty [](start = 31, length = 50)

It is not clear why presence of types necessary makes candidate unused. Is the assumption that ```member.CallingConvention``` is never unmanaged and presence of types implies that the target is unmanaged?

333fred replied:
For regular method symbols, we do not decode the extensible calling convention modopts when determining the calling convention. So, have one of 2 cases:
1. The expected convention is one of the well-known ones that maps directly to a CallKind, in which case the previous check is what we need
2. If the member's CallKind is Unmanaged and the expected is Unmanaged, we need to additionally check to make sure that the expected calling convention is not an extension convention. If we expected an extension convention that is not one of the ones that is converted to a CallKind, then those types will be in this set and there is an error. I'll rename `CallingConventionTypes` to `UnmanagedCallingConventionTypes` to hopefully make this a bit clearer.

---
In reply to: [488976731](https://github.com/dotnet/roslyn/pull/47651#discussion_r488976731) [](ancestors = 488976731)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs:
```diff
@@ -431,28 +432,116 @@ private static bool OverloadResolutionResultIsValid<TMember>(ArrayBuilder<Member
 
 #nullable enable
-        private void RemoveCallingConventionMismatches<TMember>(ArrayBuilder<MemberResolutionResult<TMember>> results, Cci.CallingConvention expectedConvention) where TMember : Symbol
+        private void RemoveCallingConventionMismatches<TMember>(ArrayBuilder<MemberResolutionResult<TMember>> results, in CallingConventionInfo expectedConvention) where TMember : Symbol
```
>RemoveCallingConventionMismatches [](start = 21, length = 33)

Where rules to detect the mismatches are specified?

333fred replied:
[https://github.com/dotnet/csharplang/blob/master/proposals/csharp-9.0/function-pointers.md#systemruntimeinteropservicesunmanagedcallersonlyattribute](https://github.com/dotnet/csharplang/blob/master/proposals/csharp-9.0/function-pointers.md#systemruntimeinteropservicesunmanagedcallersonlyattribute)

---
In reply to: [488979189](https://github.com/dotnet/roslyn/pull/47651#discussion_r488979189) [](ancestors = 488979189)

AlekseyTs replied:
Perhaps I am missing something, but It look like that section only describes how to determine calling convention from the attribute, but doesn't specify what target calling convention can be. 

---
In reply to: [490568795](https://github.com/dotnet/roslyn/pull/47651#discussion_r490568795) [](ancestors = 490568795,488979189)

333fred replied:
> doesn't specify what the target calling convention can be.

I'm not certain I understand what you mean by this. Do you mean the format for a `CallingConventionInfo`? If so, that's a standard function pointer convention, CallKind + UnmanagedCallingConventionTypes if the CallKind is Unmanaged.

---
In reply to: [490580366](https://github.com/dotnet/roslyn/pull/47651#discussion_r490580366) [](ancestors = 490580366,490568795,488979189)

AlekseyTs replied:
> I'm not certain I understand what you mean by this. 

Let me rephrase. This function filters out candidates based on ```expectedConvention```. Where are the rules specified for detecting which candidate is a valid candidate and which candidate is an invalid candidate for the given ```expectedConvention``` (whatever information it contains)?

I would expect this function to:
- calculate CallingConventionInfo for a candidate. For a candidate with UnmanagedCallingConvention attribute, content of the attribute probably will contribute to this value, let's call it ```candidateConvention```  
- match the ```candidateConvention``` from the previous step against  ```expectedConvention``` by following some rules. Candidates that failed the match will be filtered out.

The question is where the rules are specified?
Another observation is that the code is structured in a way that it appears that candidates with UnmanagedCallingConvention attribute and candidates without it are not treated uniformly, as if rules were different. If we could restructure the code to unify the handling, I think we would be able to simplify the logic to some degree. However, these are all guesses at this point, because I do not know what rules exactly we are trying to follow here. 


---
In reply to: [490606880](https://github.com/dotnet/roslyn/pull/47651#discussion_r490606880) [](ancestors = 490606880,490580366,490568795,488979189)

333fred replied:
[https://github.com/dotnet/csharplang/blob/master/proposals/csharp-9.0/function-pointers.md#function-pointer-conversions](https://github.com/dotnet/csharplang/blob/master/proposals/csharp-9.0/function-pointers.md#function-pointer-conversions)

We currently define the matching as "The calling convention of F0 is the same as the calling convention of F1." This could certainly be more precise (and I'll submit a PR to use the following rules in the definition):
* The CallKind must be the same
* If the CallKind is Unmanaged, the set of unmanaged convention types must be identical, ignoring order and duplicates.

I see your point about restructuring the checks, I'll do that.

---
In reply to: [490968551](https://github.com/dotnet/roslyn/pull/47651#discussion_r490968551) [](ancestors = 490968551,490606880,490580366,490568795,488979189)

AlekseyTs replied:
I guess the quote from [https://github.com/dotnet/csharplang/blob/master/proposals/csharp-9.0/function-pointers.md#allow-address-of-to-target-methods](https://github.com/dotnet/csharplang/blob/master/proposals/csharp-9.0/function-pointers.md#allow-address-of-to-target-methods) covers the rules: "The calling convention of M is the same as the calling convention of F. This includes both the calling convention bit, as well as any calling convention flags specified in the unmanaged identifier."

---
In reply to: [491076666](https://github.com/dotnet/roslyn/pull/47651#discussion_r491076666) [](ancestors = 491076666,490968551,490606880,490580366,490568795,488979189)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:
```diff
@@ -8250,6 +8250,19 @@ private ErrorPropertySymbol CreateErrorPropertySymbol(ImmutableArray<PropertySym
+                {
+                    foreach (var method in methodGroup.Methods)
+                    {
+                        method.ForceCompleteUnmanagedCallersOnlyAttribute();
```
>method.ForceCompleteUnmanagedCallersOnlyAttribute() [](start = 24, length = 51)

Feels wrong to call this for a PE symbol.

333fred replied:
I'm going to refactor the existing `UnmanagedCallersOnlyAttributeData` property to be a method with a `forceComplete` parameter, and do a follow up PR to make PEMethodSymbol not do a full attribute load in the implementation of that method.

---
In reply to: [489010581](https://github.com/dotnet/roslyn/pull/47651#discussion_r489010581) [](ancestors = 489010581)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:
```diff
@@ -8250,6 +8250,19 @@ private ErrorPropertySymbol CreateErrorPropertySymbol(ImmutableArray<PropertySym
+                // they are loaded to be considered for calling convention comparison. If we're in an
+                // attribute argument, then it's an error case and overload resolution will treat it as
+                // matching to suppress potentially spurious diagnostics
+                if (!InAttributeArgument && isFunctionPointerResolution)
```
>if (!InAttributeArgument && isFunctionPointerResolution) [](start = 16, length = 56)

It feels like this part belongs in Overload resolution. When it needs the information it should request it. The current approach is somewhat fragile, we have to make sure every consumer does this before OverloadResolution is invoked.

AlekseyTs commented on src/Compilers/Core/Portable/Symbols/Attributes/UnmanagedCallersOnlyAttributeData.cs:
```diff
@@ -16,6 +16,7 @@ internal sealed class UnmanagedCallersOnlyAttributeData
         internal static readonly UnmanagedCallersOnlyAttributeData Uninitialized = new UnmanagedCallersOnlyAttributeData(callingConventionTypes: ImmutableHashSet<INamedTypeSymbolInternal>.Empty, isValid: false);
         internal static readonly UnmanagedCallersOnlyAttributeData AttributePresentDataNotBound = new UnmanagedCallersOnlyAttributeData(callingConventionTypes: ImmutableHashSet<INamedTypeSymbolInternal>.Empty, isValid: false);
         private static readonly UnmanagedCallersOnlyAttributeData PlatformDefault = new UnmanagedCallersOnlyAttributeData(callingConventionTypes: ImmutableHashSet<INamedTypeSymbolInternal>.Empty, isValid: true);
+        internal static readonly UnmanagedCallersOnlyAttributeData Invalid = new UnmanagedCallersOnlyAttributeData(callingConventionTypes: ImmutableHashSet<INamedTypeSymbolInternal>.Empty, isValid: false);
```
>Invalid [](start = 67, length = 7)

What advantage do we get from using this instance?

333fred replied:
`null` is a valid case of `UnmanagedCallersOnlyAttributeData`, which means that the method has the default calling convention. I felt that the error experience would be better if we treated a mangled `UnmanagedCallersOnly` application as default unmanaged, rather than default managed.

---
In reply to: [489025845](https://github.com/dotnet/roslyn/pull/47651#discussion_r489025845) [](ancestors = 489025845)

AlekseyTs commented:
Done with review pass (iteration 1)

333fred commented:
@cston @AlekseyTs addressed feedback.

AlekseyTs commented on src/Compilers/Core/Portable/Symbols/Attributes/UnmanagedCallersOnlyAttributeData.cs:
```diff
@@ -37,7 +37,7 @@ internal static bool IsCallConvsTypedConstant(string key, in TypedConstant value
             return key == "CallConvs"
                    && value.Kind == TypedConstantKind.Array
-                   && value.Values.All(v => v.Kind == TypedConstantKind.Type);
+                   && (value.Values.IsDefaultOrEmpty || value.Values.All(v => v.Kind == TypedConstantKind.Type));
```
>v.Kind == TypedConstantKind.Type [](start = 78, length = 32)

Do we want to treat null values in the array as illegal? Is this check going to do that?

333fred replied:
Null values in the array are illegal, [https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Portable/Symbols/MethodSymbol.cs#L1006](https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Portable/Symbols/MethodSymbol.cs#L1006) checks it. It's tested here: [https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs#L8051](https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs#L8051)

---
In reply to: [491007732](https://github.com/dotnet/roslyn/pull/47651#discussion_r491007732) [](ancestors = 491007732)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/MethodSymbol.cs:
```diff
@@ -104,30 +104,14 @@ public virtual bool IsGenericMethod
-            Debug.Assert(!ReferenceEquals(UnmanagedCallersOnlyAttributeData, UnmanagedCallersOnlyAttributeData.Uninitialized));
-            Debug.Assert(!ReferenceEquals(UnmanagedCallersOnlyAttributeData, UnmanagedCallersOnlyAttributeData.AttributePresentDataNotBound));
-        }
+        /// Returns the UnmanagedCallersOnlyAttribute data for this method, if there is any. If the data has not yet been
```
>UnmanagedCallersOnlyAttribute data [](start = 24, length = 34)

Consider using cref for the type

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbolWithAttributes.cs:
```diff
@@ -395,55 +395,65 @@ internal sealed override ObsoleteAttributeData ObsoleteAttributeData
-#if DEBUG // Can remove ifdefs and replace with Conditional after https://github.com/dotnet/roslyn/issues/47463 is fixed
-                    verifyDataConsistent((CommonMethodEarlyWellKnownAttributeData?)lazyCustomAttributesBag.EarlyDecodedWellKnownAttributeData, data);
+            var lazyCustomAttributesBag = _lazyCustomAttributesBag;
+            if (lazyCustomAttributesBag is null || !lazyCustomAttributesBag.IsEarlyDecodedWellKnownAttributeDataComputed)
```
>!lazyCustomAttributesBag.IsEarlyDecodedWellKnownAttributeDataComputed [](start = 51, length = 69)

Is it possible that IsDecodedWellKnownAttributeDataComputed returns true, but IsEarlyDecodedWellKnownAttributeDataComputed returns false? If not, an assert somewhere (perhaps in IsEarlyDecodedWellKnownAttributeDataComputed) could be helpful.

333fred replied:
I don't believe so, because the code that sets the normal data is the same path that sets the early data, it just bails after setting the early data if early is all that's requested: [http://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp/Symbols/Symbol_Attributes.cs,262](http://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp/Symbols/Symbol_Attributes.cs,262). I can add an assert to the two getters though.

---
In reply to: [491021696](https://github.com/dotnet/roslyn/pull/47651#discussion_r491021696) [](ancestors = 491021696)

333fred commented:
@AlekseyTs addressed feedback. I restructured the overload resolution comparison to be simpler and explicitly spell out the rules involved, I believe it's what you were looking for.

AlekseyTs commented:
>         public void UnmanagedCallersOnlyCallConvNull()

I think we would want to strengthen this test in a sense that we should try to use an address of M and see if the behavior is what we expect. Should also cover scenario of the same attribute coming from metadata.

---
Refers to: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:8051 in 0c95f73. [](commit_id = 0c95f73b7867db3a8a2e96324c55eb6b7e9e67ed, deletion_comment = False)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs:
```diff
@@ -431,28 +432,142 @@ private static bool OverloadResolutionResultIsValid<TMember>(ArrayBuilder<Member
+                    }
+                    else
+                    {
+                        // There's data from an UnmanagedCallersOnlyAttribute present, with takes precedence over the
```
>with [](start = 87, length = 4)

Typo?

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs:
```diff
@@ -431,28 +432,142 @@ private static bool OverloadResolutionResultIsValid<TMember>(ArrayBuilder<Member
+                        //   the CallKind is treated as Unmanaged, with the union of the types specified treated as calling convention types.
+
+                        var unmanagedCallingConventionTypes = unmanagedCallersOnlyData.CallingConventionTypes;
+                        Debug.Assert(unmanagedCallingConventionTypes.All(u => FunctionPointerTypeSymbol.IsCallingConventionModifier((NamedTypeSymbol)u)));
```
>u [](start = 149, length = 1)

Could we have a null in the set? It looks like ```IsCallingConventionModifier``` isn't ready to deal with nulls.

333fred replied:
No. In source, a `null` won't get added. And from metadata, seeing a `null` will cause us to treat the data as invalid and issue a use-site diagnostic on the use of `M`.

---
In reply to: [491113023](https://github.com/dotnet/roslyn/pull/47651#discussion_r491113023) [](ancestors = 491113023)

333fred replied:
Added an additional test showing metadata behavior around null.

---
In reply to: [491121238](https://github.com/dotnet/roslyn/pull/47651#discussion_r491121238) [](ancestors = 491121238,491113023)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs:
```diff
@@ -431,28 +432,142 @@ private static bool OverloadResolutionResultIsValid<TMember>(ArrayBuilder<Member
+                    //    and duplicates. We already have both sets in a HashSet, so we can just ensure they're the same length and
+                    //    that everything from one set is in the other set.
+
+                    if (!actualCallKind.IsCallingConvention(expectedConvention.CallKind))
```
>!actualCallKind.IsCallingConvention(expectedConvention.CallKind) [](start = 24, length = 64)

It looks like ConversionBase.HasImplicitPointerConversion simply uses equality check. Is there a reason to differ?

333fred replied:
Yeah, we could change this. I'll add an assert that the expected convention has no unknown bits in it.

---
In reply to: [491116706](https://github.com/dotnet/roslyn/pull/47651#discussion_r491116706) [](ancestors = 491116706)

333fred replied:
@AlekseyTs update: `member.CallingConvention` here can include both a CallKind and a possible calling convention attribute (such as Generic/HasThis/ExplicitThis). HasImplicitPointerConversion doesn't need to worry about these, because those are both function pointer types and won't have a convention attribute. However, we do have to worry about them in this case because a regular method may have them, and `IsCallingConvention` will do the correct comparison here.

---
In reply to: [491132506](https://github.com/dotnet/roslyn/pull/47651#discussion_r491132506) [](ancestors = 491132506,491116706)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs:
```diff
@@ -431,28 +432,142 @@ private static bool OverloadResolutionResultIsValid<TMember>(ArrayBuilder<Member
+
+                    if (expectedConvention.CallKind.IsCallingConvention(Cci.CallingConvention.Unmanaged))
+                    {
+                        switch (expectedConvention.UnmanagedCallingConventionTypes, actualUnmanagedCallingConvetionTypes)
```
>switch (expectedConvention.UnmanagedCallingConventionTypes, actualUnmanagedCallingConvetionTypes) [](start = 24, length = 97)

It looks like ConversionBase.HasImplicitPointerConversion simply uses SetEquals helper. Is there a reason to use custom implementation here? It feels like the calling convention comparison logic should be extracted into a helper and shared between ConversionsBase and OverloadResolution.

333fred replied:
I forgot about that method, thanks.

---
In reply to: [491117561](https://github.com/dotnet/roslyn/pull/47651#discussion_r491117561) [](ancestors = 491117561)

333fred replied:
Oh, actually, no, we can't use that method here. We have a set of `CustomModifier`s and a set of `INamedTypeSymbolInternal`, and I didn't feel that doing an allocation to get custom modifiers into a state that could be set equaled with a hashset was worth it.

---
In reply to: [491121777](https://github.com/dotnet/roslyn/pull/47651#discussion_r491121777) [](ancestors = 491121777,491117561)

AlekseyTs replied:
> We have a set of CustomModifiers and a set of INamedTypeSymbolInternal

I see.

---
In reply to: [491122866](https://github.com/dotnet/roslyn/pull/47651#discussion_r491122866) [](ancestors = 491122866,491121777,491117561)

AlekseyTs commented:
Done with review pass (iteration 6)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [695018418](https://github.com/dotnet/roslyn/pull/47651#issuecomment-695018418) [](ancestors = 695018418)

AlekseyTs reviewed (approved):
LGTM (iteration 7)

333fred commented:
@cston for another review please.

cston commented on src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs:
```diff
@@ -431,28 +432,137 @@ private static bool OverloadResolutionResultIsValid<TMember>(ArrayBuilder<Member
+            {
+                // We're at a location where the unmanaged data might not yet been bound. This cannot be valid code
+                // anyway, as attribute arguments can't be method references, so we'll just assume that the conventions
+                // match, as there will be other errors that superscede these anyway
```
>superscede [](start = 61, length = 10)

Typo.

cston commented on src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs:
```diff
@@ -431,28 +432,137 @@ private static bool OverloadResolutionResultIsValid<TMember>(ArrayBuilder<Member
+                                 && !ReferenceEquals(unmanagedCallersOnlyData, UnmanagedCallersOnlyAttributeData.Uninitialized));
+
+                    Cci.CallingConvention actualCallKind;
+                    ImmutableHashSet<INamedTypeSymbolInternal> actualUnmanagedCallingConvetionTypes;
```
>Convetion [](start = 85, length = 9)

Typo.

cston commented on src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs:
```diff
@@ -431,28 +432,137 @@ private static bool OverloadResolutionResultIsValid<TMember>(ArrayBuilder<Member
+                            case ({ Count: var expectedCount }, { Count: var actualCount }) when expectedCount != actualCount:
+                                results[i] = makeWrongCallingConvention(result);
+                                continue;
+                        }
```
This is a little difficult to read. Could it be simplified to the following?
```C#
if (expectedConvention.UnmanagedCallingConventionTypes.Count != actualUnmanagedCallingConvetionTypes.Count)
{
    results[i] = makeWrongCallingConvention(result);
}
```

cston commented on src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs:
```diff
@@ -431,28 +432,137 @@ private static bool OverloadResolutionResultIsValid<TMember>(ArrayBuilder<Member
+                                continue;
+                        }
+
+                        foreach (var expectedModifier in expectedConvention.UnmanagedCallingConventionTypes)
```
>foreach (var expectedModifier in expectedConvention.UnmanagedCallingConventionTypes) [](start = 24, length = 84)

Consider using:
```C#
if (!actualUnmanagedCallingConvetionTypes.SetEquals(expectedConvention.UnmanagedCallingConventionTypes))
{
    results[i] = makeWrongCallingConvention(result);
}
```

333fred replied:
We can't use `SetEquals`, these collections are of different types.

cston commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -9812,6 +9905,312 @@ public UnmanagedCallersOnlyAttribute()
+                //     [Attr(PropCdecl = &M1)]
+                Diagnostic(ErrorCode.ERR_BadNamedAttributeArgumentType, "PropCdecl", isSuppressed: false).WithArguments("PropCdecl").WithLocation(22, 11)
+            );
+        }
```
>} [](start = 8, length = 1)

Consider testing:
```C#
using System;
using System.Runtime.InteropServices;
[AttributeUsage(AttributeTargets.Method, AllowMultiple = true)]
unsafe class Attr : Attribute
{
    public Attr(delegate*<void> d) {}
}
unsafe class C
{
    [UnmanagedCallersOnly]
    [Attr(&M1)]
    static void M1()
    {
    }
}
```

cston replied:
Consider testing:
```C#
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;
public unsafe class C
{
    [UnmanagedCallersOnly(CallConvs = new[] { typeof(CallConvFastcall) })]
    static void F(int i = G(&F)) { }
    static int G(delegate*unmanaged[Fastcall]<int, void> d) => 0;
}
```

---
In reply to: [491271217](https://github.com/dotnet/roslyn/pull/47651#discussion_r491271217) [](ancestors = 491271217)

333fred commented:
@cston addressed feedback.

cston commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -9745,6 +9838,155 @@ class C
+using System;
+using System.Runtime.InteropServices;
+[AttributeUsage(AttributeTargets.Method, AllowMultiple = true)]
+unsafe class Attr : Attribute
```
>unsafe [](start = 0, length = 6)

Minor: `unsafe` is not needed.

cston commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -9745,6 +9838,155 @@ class C
+    [Attr(F())]
+    static int F()
+    {
+    }
```
Minor: Consider returning a value to remove the unnecessary error.
